misc/package/Data Files/MWSE/core/lib/logger/init.lua

```diff
@@ -0,0 +1,681 @@
+    format = true,
+}
+
+-- metatable used by log objects
+local logMetatable = {
+    -- gonna override this later so that it's exactly equal to `Logger.debug`. this is needed for line number stuff to work properly
+    __call = function(self, ...) self:debug(...) end,
+    __index = Logger,
+    __newindex = function(self, k, v)
+        if k == "logLevel" then
+            -- this will upgrade everybody's log level
+            self:setLogLevel(v)
+        else
+            rawset(self, k, v)
+        end
+    end,
+    
+    __tostring = function(self)
+        return fmt("Logger(modName: %q, moduleName: %s, modDir: %q, level: %i (%s))",
+            self.modName, self.moduleName and fmt("%q", self.moduleName), self.modDir, self.level, self:getLevelStr()
+        )
+    end,
+}
+
+-- these will be checked against `newInfo.source` to see whether we've gone too far up the stack when trying to get the filepath
+-- of the file constructing this logger
+local badFilePaths = {
+    [string.lower("@Data Files\\MWSE\\core\\initialize.lua")] = true,
+    [string.lower("@Data Files\\MWSE\\core\\startLuaMods.lua")] = true,
+    [string.lower("@.\\Data Files\\MWSE\\core\\lib\\logger\\init.lua")] = true,
+    [string.lower("=[C]")] = true,
+}
+
+
+---@return string? modName, string? moduleName, string? modDir
+local function getModInfoFromSource()
+
+    -- =========================================================================
+    -- generate relevant mod information
+    -- =========================================================================
+
+    
+
+    -- iterate up a few times to get the correct path when people use logger factories
+    -- i.e., we want to handle the case where a mod constructs a logger by 
+    -- calling another function that constructs that logger
+
+    local info, newInfo
+    local i = 2
+    repeat
+        i = i + 1
+        info = newInfo
+        newInfo = debug.getinfo(i, "S")
+    until not newInfo or badFilePaths[newInfo.source:lower()]
+
+    local src = info and info.source
+    if not src then return end
+
+
+    -- parts of the path without "@^\Data Files\MWSE\mods\"
+    local luaParts = table.filterarray(src:split("\\/"), function(i) return i >= 5 end)
+    
+    -- this happens if the logger is being constructed in a tail-recursive way
+    if #luaParts == 0 then return end
+
+    local hasAuthorName = false
+
+
+    -- first check for metadata when `luaParts[1]` is the mod author directory
+    local metadata = tes3.getLuaModMetadata(luaParts[1] .. "." .. luaParts[2]) ---@type MWSE.Metadata?
+    if metadata then
+        hasAuthorName = true
+    else
+        -- then check for metadata when `luaParts[1]` is the mod root directory
+        metadata = tes3.getLuaModMetadata(luaParts[1])
+    end
+    if metadata then
+        hasAuthorName = hasAuthorName or false
+    else
+        local oneDirUp = table.concat({"Data Files", "MWSE", "mods", luaParts[1]}, "\\")
+        local lfs = require("lfs")
```

Comment:
`lfs` library is already in global namespace.

Reply:
I could be misremembering, but I think I added this a few months ago when there were some weird crashes/errors happening from the logger being loaded before `core/initialize.lua` finished running. (I think because some things like the `DependencyManager` wanted a logger.)

But I tested this just now and it seems like things mostly work fine without the `lfs = require("lfs")` code. Although I'm still scared to delete it. I might be thinking of something else though.